### PR TITLE
ARROW-14522: [C++] Fix validation of ExtensionType with null storage type

### PR DESCRIPTION
**Pull request by jorisvandenbossche**

### CODE REVIEW

File: cpp/src/arrow/array/validate.cc [hunk 1]
```diff
@@ -294,6 +294,16 @@ struct ValidateArrayImpl {
 
 }  // namespace
 
+bool IsNullType(const DataType& type) {
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/array/validate.cc:297]: Is such a helper function the best approach? And if so, is it fine to just put it here?

Comment (pitrou) [reply cpp/src/arrow/array/validate.cc:297]: We could perhaps expose a `storage_type_id` method in `DataType`.

Comment (jorisvandenbossche) [reply cpp/src/arrow/array/validate.cc:297]: Alternative could be to check the type's layout and check if `type.layout().buffers[0].kind == DataTypeLayout::ALWAYS_NULL` ?

Comment (jorisvandenbossche) [reply cpp/src/arrow/array/validate.cc:297]: > We could perhaps expose a storage_type_id method in DataType.

And which returns itself if it's not an extension type?

Comment (pitrou) [reply cpp/src/arrow/array/validate.cc:297]: Right. cc @bkietz for opinions.

File: python/pyarrow/tests/test_extension_type.py [hunk 2]
```diff
@@ -76,6 +76,19 @@ def __reduce__(self):
         return MyListType, (self.storage_type,)
 
 
+class AnnotatedType(pa.PyExtensionType):
+    """
+    Generic extension type that can store any storate type.
```
Comment (lidavidm) [inline python/pyarrow/tests/test_extension_type.py:81]:
```suggestion
    Generic extension type that can store any storage type.
```

**Comment by pitrou** [comment]
@jorisvandenbossche Do you want to rebase and update this?

**Comment by pitrou** [comment]
Ping @jorisvandenbossche

File: cpp/src/arrow/array/validate.cc (hunk 1 above)
Comment (jorisvandenbossche) [reply cpp/src/arrow/array/validate.cc]: I tried it now with a `DataType::storage_id()`

**Review by pitrou — APPROVED** [review]
LGTM. Thank you @jorisvandenbossche !

**Pull request closed without merging**